### DISCUSSION TITLE
Use caplog in autouse fixture

### DISCUSSION THREAD

**Discussion by eflavio** [pytest-dev/pytest#11986] [Q&A]
Hello. I'd like to wrap all my tests with a fixture where the logs are checked for error messages. I tried this:

```
@pytest.fixture(autouse=True)
def assert_no_log_error(caplog):
    yield
    assert "ERROR" not in caplog.text
```

But caplog.text is always empty. I assume, caplog is cleared after the test and before the fixture actually checks the logs.
How can I make this work?

**Comment by eflavio** [comment]
I asked the question in Stackoverflow: https://stackoverflow.com/questions/78274221/use-caplog-in-autouse-fixture-in-pytest

**Comment by dtczest** [comment]
Having the same issue :( I would have thought the caplog wouldn't be cleared until after the test is torn down.

**Comment by eflavio** [comment]
I go it to work: https://stackoverflow.com/a/78276016/24008810

```
import logging
from _pytest.logging import LogCaptureFixture


@pytest.fixture(autouse=True)
def assert_no_log_error(caplog: LogCaptureFixture):
    yield
    for record in caplog.get_records('call'):
        assert record.levelno < logging.ERROR
```

**Reply by dtczest** [reply]
Ahh I see what you did. Unfortunately I need the *text* because I'm using a formatter. They don't seem to stash that information :(

**Reply by eflavio** [reply]
Try `record.msg` instead of `record.levelno`

**Reply by dtczest** [reply]
unfortunately the formatted stream is not stored on the record. That is just the message that's passed into the log call. I can't format it after the call because my formatting relies on values available at runtime.

**Reply by eflavio** [reply]
I don't see where you would have a problem. Can you describe your problem with a little more context/example?

**Reply by eflavio** [reply]
You can use a `logging.Formatter` instance to get the formatted message from a `logging.LogRecord`.

**Reply by dtczest** [reply]
Right, except the formatting code takes uses some context from the running code in order to include some extra data. That context would have already been deleted.
For example, we set a context variable that has an ID in it. The formatter looks at that context variable and adds it into the log message. After the code context is complete, the variable is cleared, so calling the formatter after the fact will NOT include the context variable.

I appreciate the help, though! Thanks.
